### DISCUSSION TITLE
Custom DB operations on record insert

### DISCUSSION THREAD

**Discussion by whmeitzler** [pocketbase/pocketbase#5136] [Q&A] [answered]
Hello!
I have custom behavior on certain collections to make them behave as a ringbuffer/circular queue. 
In essence I intercept create operations and execute custom SQL, converting a CREATE to an UPDATE (at the correct record location). 

At the end of the hook, I return a hooks.StopPropagation error. This disables the default SQL INSERT from executing, which is great. 

However, if I want to be able to subscribe to these collections, it seems that there is a hook after the default SQL INSERT that feeds the realtime broker. That is, I have a choice between a working ringbuffer or the ability to subscribe to these collections. 

Is there a way to call this function directly from the onCreate hook? Or is there a more detailed error I can pass that skips only the DB operation?

Sample hook:
```go

		onCreateHookId := app.OnModelBeforeCreate(ring.coll.Name).
			Add(func(e *core.ModelEvent) error {

				if vals, ok := e.Model.(models.ColumnValueMapper); !ok {
				} else {
					params := vals.ColumnValueMap()
					//Convert a create into an update
					delete(params, "id")
					nextSeq := ring.HighestSeq + 1
					params[sequenceColumnName] = nextSeq
					nextRowId := ((ring.HighestSeq) % (ring.Size)) + 1
					if ring.RecordCount < ring.Size {
						if _, err := app.Dao().
							NonconcurrentDB().
							Insert(
								ring.coll.Name,
								params,
							).Execute(); err != nil {
							return err
						}
						ring.RecordCount++

					} else {
						q := app.Dao().
							NonconcurrentDB().
							Update(
								ring.coll.Name,
								params,
								dbx.HashExp{
									"rowid": nextRowId,
								})
						q.ExecLogFunc = func(ctx context.Context, t time.Duration, sql string, result sql.Result, err error) {
							fmt.Println(sql)
						}

						if _, err := q.Execute(); err != nil {
							return err
						}
					}
					ring.HighestSeq = nextSeq
					ring.onInsert.Trigger(Event{
						Ringbuffer: ring,
						RowId:      nextRowId,
						Seq:        nextSeq,
						Model:      e.Model,
					})
				}
                               //return nil //realtime works, but I get the default SQL operation too 
				return hook.StopPropagation //SQL behaves correctly, but realtime is not triggered
			})
```

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not entirely sure that I understand what the above code is supposed to do but the realtime create/update events are fired in the after hook which is triggered as finalizer of the before hook. In your case since you want to prevent the default create behavior you can try to add before the StopPropagation call something like:

```go
if err := app.OnModelAfterCreate().Trigger(e); err != nil {
	return err
}
```

_Simpler hooks before-after handling and more advanced priority management is implemented as part of the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355)._

**Reply by ganigeorgiev** [reply]
> In essence I intercept create operations and execute custom SQL, converting a CREATE to an UPDATE (at the correct record location).

_Side-not: Instead of changing the type of the operation it may be easier to simply perform a delete of the last N records on successful create._

**Comment by whmeitzler** [comment]
Triggering the OnAfterModelCreate solves my problem for now. 
Thanks, as always, for your support. I'm excited to see the fruits of your refactoring efforts!
